src/main/java/org/sonarsource/sonarlint/ls/connected/sync/ServerSynchronizer.java

```diff
@@ -0,0 +1,178 @@
+        new MessageParams(MessageType.Error, "Binding update failed for the following connection(s): " + connections + ". Look at the SonarLint output for details."));
+    }
+  }
+
+  private Set<String> tryUpdateConnectionsAndBoundProjectStorages(Map<String, Set<String>> projectKeyByConnectionIdsToUpdate, ProgressFacade progress) {
+    var failedConnectionIds = new LinkedHashSet<String>();
+    projectKeyByConnectionIdsToUpdate.forEach(
+      (connectionId, projectKeys) -> tryUpdateConnectionAndBoundProjectsStorages(projectKeyByConnectionIdsToUpdate, progress, failedConnectionIds, connectionId, projectKeys));
+    return failedConnectionIds;
+  }
+
+  private void tryUpdateConnectionAndBoundProjectsStorages(Map<String, Set<String>> projectKeyByConnectionIdsToUpdate, ProgressFacade progress,
+    Set<String> failedConnectionIds, String connectionId, Set<String> projectKeys) {
+    progress.doInSubProgress(connectionId, 1.0f / projectKeyByConnectionIdsToUpdate.size(), subProgress -> {
+      var endpointParamsAndHttpClient = bindingManager.getServerConfigurationFor(connectionId);
+      if (endpointParamsAndHttpClient == null) {
+        failedConnectionIds.add(connectionId);
+        return;
+      }
+      var engineOpt = bindingManager.getOrCreateConnectedEngine(connectionId);
+      if (engineOpt.isEmpty()) {
+        failedConnectionIds.add(connectionId);
+        return;
+      }
+      subProgress.doInSubProgress("Update projects storages", 0.5f, s -> tryUpdateBoundProjectsStorage(
+        projectKeys, endpointParamsAndHttpClient, engineOpt.get(), s));
+      subProgress.doInSubProgress("Sync projects storages", 0.5f, s -> syncOneEngine(
+        connectionId, projectKeys, engineOpt.get(), s));
+    });
+  }
+
+  private static void tryUpdateBoundProjectsStorage(Set<String> projectKeys, ServerConnectionSettings.EndpointParamsAndHttpClient endpointParamsAndHttpClient,
+    ConnectedSonarLintEngine engine,
+    ProgressFacade progress) {
+    projectKeys.forEach(projectKey -> progress.doInSubProgress(projectKey, 1.0f / projectKey.length(), subProgress -> {
+      try {
+        engine.updateProject(endpointParamsAndHttpClient.getEndpointParams(), endpointParamsAndHttpClient.getHttpClient(), projectKey, subProgress.asCoreMonitor());
+      } catch (CanceledException e) {
+        throw e;
+      } catch (Exception updateFailed) {
+        LOG.error("Binding update failed for project key '{}'", projectKey, updateFailed);
+      }
+    }));
+  }
+
+  private void syncStorage() {
+    var projectKeyPerConnectionId = new HashMap<String, Set<String>>();
+    bindingManager.forEachBoundFolder((folder, settings) -> {
+      var connectionId = requireNonNull(settings.getConnectionId());
+      var projectKey = requireNonNull(settings.getProjectKey());
+      projectKeyPerConnectionId
+        .computeIfAbsent(connectionId, k -> new HashSet<>())
+        .add(projectKey);
+    });
+    if (!projectKeyPerConnectionId.isEmpty()) {
```

Comment:
Maybe flip the condition and return if empty?

Reply:
The code has moved and the method is a bit shorter. I like to exit early if the method is long, here I think it's ok